### ISSUE TITLE
Hugo compiles non deterministic output when using custom output format

### ISSUE THREAD

**Issue by noreff** [gohugoio/hugo#8110]
### What version of Hugo are you using (`hugo version`)?
<pre>
Hugo Static Site Generator v0.80.0-792EF0F4 darwin/amd64 BuildDate: 2020-12-31T13:36:47Z
</pre>

### Does this issue reproduce with the latest release?
Yes

### Details

I'm using Hugo to create php files. And it was working just fine, but starting from version 0.63.0 Hugo start occasionally producing different output when running `hugo` (without changing any files at all). Version 0.62.2 works without any issues.

See https://github.com/noreff/hugo-non-deterministic-output-bug/raw/master/hugo_bug.gif

I created a minimal reproducible example and upload it to github - https://github.com/noreff/hugo-non-deterministic-output-bug.

This starts happening in this release - https://github.com/gohugoio/hugo/releases/tag/v0.63.0.
More specific - after this commit - https://github.com/gohugoio/hugo/commit/c6d650c8c8b22fdc7ddedc1e42a3ca698e1390d6 (tpl/tplimpl: Rework template management to get rid of concurrency issues) which is quite huge. I tried to find the reason myself, but didn't succeed yet. I will be very grateful for any help.

**Comment by moorereason** [comment]
@noreff,
I removed the link to the 9MB gif in the OP.  It's too big and not that helpful anyway.

@bep,
I can replicate the issue with Hugo master.  Periodically, Hugo renders foo's content for the bar page. I've [submitted a PR](https://github.com/noreff/hugo-non-deterministic-output-bug/pull/1) to the demo repo to trim it down to the bare minimum ([my branch](https://github.com/moorereason/hugo-non-deterministic-output-bug/tree/simplify)).  It looks like the bug is in the baseof functionality.  If I remove the blocks from `layouts/bar/single.php`, I haven't seen it generate the wrong content after 5000 iterations.  With blocks, it will generate the wrong content in under 500 iterations or so.  In my PR/branch, I've included a `run.sh` script to build the site repeatedly until the sha1sum for `public/bar/index.php` is different.

**Comment by moorereason** [comment]
I can confirm that the issue appears to start with https://github.com/gohugoio/hugo/commit/c6d650c8c8b22fdc7ddedc1e42a3ca698e1390d6.

Also, I'm running go1.16beta1 linux/amd64.

**Comment by moorereason** [comment]
It looks like a race condition with text templates.  If `isPlainText = false` in the outputFormats definition, there's no race detected.

To reproduce, build Hugo with the race detector enabled:
```
$ go install -ldflags "-X github.com/gohugoio/hugo/common/hugo.commitHash=4fdec67b" -race github.com/gohugoio/hugo
```
Build [the simplified demo site](https://github.com/moorereason/hugo-non-deterministic-output-bug/tree/simplify):
```
$ hugo version
Hugo Static Site Generator v0.81.0-DEV-4FDEC67B linux/amd64 BuildDate: unknown

$ hugo 2>&1 | sed 's/home.cmoore.//g'
Start building sites …
==================
WARNING: DATA RACE
Write at 0x00c000981c20 by goroutine 52:
  runtime.mapassign_faststr()
      /go/src/runtime/map_faststr.go:202 +0x0
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).associate()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:227 +0x1a8
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).AddParseTree()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:133 +0x2cc
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).Parse()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:208 +0xc88
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).applyBaseTemplate()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:595 +0x484
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).findLayout()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:402 +0x7f7
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).LookupLayout()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:320 +0x357
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateExec).LookupLayout()
      <autogenerated>:1 +0x177
  github.com/gohugoio/hugo/hugolib.(*pageState).resolveTemplate()
      /src/github.com/gohugoio/hugo/hugolib/page.go:489 +0x354
  github.com/gohugoio/hugo/hugolib.pageRenderer()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:124 +0x1c4

Previous read at 0x00c000981c20 by goroutine 53:
  runtime.mapaccess1_faststr()
      /go/src/runtime/map_faststr.go:12 +0x0
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walkTemplate()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:399 +0x13b
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:268 +0x2e4
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:263 +0x1a4
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).executeWithState()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:93 +0x150
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*executer).Execute()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:79 +0x29c
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateExec).Execute()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:226 +0x1ca
  github.com/gohugoio/hugo/hugolib.(*Site).renderForTemplate()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1747 +0xec
  github.com/gohugoio/hugo/hugolib.(*Site).renderAndWritePage()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1677 +0x3ec
  github.com/gohugoio/hugo/hugolib.pageRenderer()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:137 +0x458

Goroutine 52 (running) created at:
  github.com/gohugoio/hugo/hugolib.(*Site).renderPages()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:73 +0x1a4
  github.com/gohugoio/hugo/hugolib.(*Site).render()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1221 +0x90
  github.com/gohugoio/hugo/hugolib.(*HugoSites).render()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:320 +0x987
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build.func5()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:149 +0x57
  runtime/trace.WithRegion()
      /go/src/runtime/trace/annotation.go:141 +0x150
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:151 +0xadb
  github.com/gohugoio/hugo/commands.(*commandeer).buildSites()
      /src/github.com/gohugoio/hugo/commands/hugo.go:722 +0x12d
  github.com/gohugoio/hugo/commands.(*commandeer).fullBuild.func3()
      /src/github.com/gohugoio/hugo/commands/hugo.go:302 +0x138
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /pkg/mod/golang.org/x/sync@v0.0.0-20200317015054-43a5402ce75a/errgroup/errgroup.go:57 +0x94

Goroutine 53 (finished) created at:
  github.com/gohugoio/hugo/hugolib.(*Site).renderPages()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:73 +0x1a4
  github.com/gohugoio/hugo/hugolib.(*Site).render()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1221 +0x90
  github.com/gohugoio/hugo/hugolib.(*HugoSites).render()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:320 +0x987
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build.func5()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:149 +0x57
  runtime/trace.WithRegion()
      /go/src/runtime/trace/annotation.go:141 +0x150
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:151 +0xadb
  github.com/gohugoio/hugo/commands.(*commandeer).buildSites()
      /src/github.com/gohugoio/hugo/commands/hugo.go:722 +0x12d
  github.com/gohugoio/hugo/commands.(*commandeer).fullBuild.func3()
      /src/github.com/gohugoio/hugo/commands/hugo.go:302 +0x138
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /pkg/mod/golang.org/x/sync@v0.0.0-20200317015054-43a5402ce75a/errgroup/errgroup.go:57 +0x94
==================
==================
WARNING: DATA RACE
Write at 0x00c000944638 by goroutine 52:
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).associate()                                                                                                             [39/4550]
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:227 +0x1bd
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).AddParseTree()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:133 +0x2cc
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).Parse()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/template.go:208 +0xc88
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).applyBaseTemplate()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:601 +0x348
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).findLayout()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:402 +0x7f7
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateHandler).LookupLayout()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:320 +0x357
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateExec).LookupLayout()
      <autogenerated>:1 +0x177
  github.com/gohugoio/hugo/hugolib.(*pageState).resolveTemplate()
      /src/github.com/gohugoio/hugo/hugolib/page.go:489 +0x354
  github.com/gohugoio/hugo/hugolib.pageRenderer()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:124 +0x1c4

Previous read at 0x00c000944638 by goroutine 53:
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walkTemplate()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:399 +0x14e
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:268 +0x2e4
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:263 +0x1a4
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).executeWithState()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:93 +0x150
  github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*executer).Execute()
      /src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:79 +0x29c
  github.com/gohugoio/hugo/tpl/tplimpl.(*templateExec).Execute()
      /src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:226 +0x1ca
  github.com/gohugoio/hugo/hugolib.(*Site).renderForTemplate()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1747 +0xec
  github.com/gohugoio/hugo/hugolib.(*Site).renderAndWritePage()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1677 +0x3ec
  github.com/gohugoio/hugo/hugolib.pageRenderer()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:137 +0x458

Goroutine 52 (running) created at:
  github.com/gohugoio/hugo/hugolib.(*Site).renderPages()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:73 +0x1a4
  github.com/gohugoio/hugo/hugolib.(*Site).render()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1221 +0x90
  github.com/gohugoio/hugo/hugolib.(*HugoSites).render()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:320 +0x987
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build.func5()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:149 +0x57
  runtime/trace.WithRegion()
      /go/src/runtime/trace/annotation.go:141 +0x150
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:151 +0xadb
  github.com/gohugoio/hugo/commands.(*commandeer).buildSites()
      /src/github.com/gohugoio/hugo/commands/hugo.go:722 +0x12d
  github.com/gohugoio/hugo/commands.(*commandeer).fullBuild.func3()
      /src/github.com/gohugoio/hugo/commands/hugo.go:302 +0x138
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /pkg/mod/golang.org/x/sync@v0.0.0-20200317015054-43a5402ce75a/errgroup/errgroup.go:57 +0x94

Goroutine 53 (finished) created at:
  github.com/gohugoio/hugo/hugolib.(*Site).renderPages()
      /src/github.com/gohugoio/hugo/hugolib/site_render.go:73 +0x1a4
  github.com/gohugoio/hugo/hugolib.(*Site).render()
      /src/github.com/gohugoio/hugo/hugolib/site.go:1221 +0x90
  github.com/gohugoio/hugo/hugolib.(*HugoSites).render()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:320 +0x987
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build.func5()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:149 +0x57
  runtime/trace.WithRegion()
      /go/src/runtime/trace/annotation.go:141 +0x150
  github.com/gohugoio/hugo/hugolib.(*HugoSites).Build()
      /src/github.com/gohugoio/hugo/hugolib/hugo_sites_build.go:151 +0xadb
  github.com/gohugoio/hugo/commands.(*commandeer).buildSites()
      /src/github.com/gohugoio/hugo/commands/hugo.go:722 +0x12d
  github.com/gohugoio/hugo/commands.(*commandeer).fullBuild.func3()
      /src/github.com/gohugoio/hugo/commands/hugo.go:302 +0x138
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /pkg/mod/golang.org/x/sync@v0.0.0-20200317015054-43a5402ce75a/errgroup/errgroup.go:57 +0x94
==================

                   | EN
-------------------+-----
  Pages            |  2
  Paginator pages  |  0
  Non-page files   |  0
  Static files     |  0
  Processed images |  0
  Aliases          |  0
  Sitemaps         |  0
  Cleaned          |  0

Total in 46 ms
Found 2 data race(s)
```

**Comment by noreff** [comment]
I forget to mention, that occasionally (quite rare) I also was receiving an error message that looked like this:

```
Start building sites … 
fatal error: concurrent map read and map write

goroutine 99 [running]:
runtime.throw(0x5d696fe, 0x21)
        /usr/local/Cellar/go/1.15.4/libexec/src/runtime/panic.go:1116 +0x72 fp=0xc0008bf4b0 sp=0xc0008bf480 pc=0x403c492
runtime.mapaccess1_faststr(0x5a133e0, 0xc000624870, 0xc0002520ff, 0x4, 0x8b6)
        /usr/local/Cellar/go/1.15.4/libexec/src/runtime/map_faststr.go:21 +0x465 fp=0xc0008bf520 sp=0xc0008bf4b0 pc=0x40197a5
github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walkTemplate(0xc0008bf7e8, 0x5cd5e40, 0xc000ae3f50, 0x16, 0xc0008b8e00)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:399 +0x8e fp=0xc0008bf608 sp=0xc0008bf520 pc=0x4a7db8e
github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk(0xc0008bf7e8, 0x5cd5e40, 0xc000ae3f50, 0x16, 0x61ea320, 0xc0008b8e00)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:268 +0x1da fp=0xc0008bf690 sp=0xc0008bf608 pc=0x4a7cd3a
github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*state).walk(0xc0008bf7e8, 0x5cd5e40, 0xc000ae3f50, 0x16, 0x61ea0e0, 0xc0008c44e0)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/exec.go:263 +0x11e fp=0xc0008bf718 sp=0xc0008bf690 pc=0x4a7cc7e
github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*Template).executeWithState(0xc0008b8bc0, 0xc0008bf7e8, 0x5cd5e40, 0xc000ae3f50, 0x16, 0x0, 0x0)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:93 +0xb7 fp=0xc0008bf780 sp=0xc0008bf718 pc=0x4a8a617
github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate.(*executer).Execute(0xc000a8f930, 0xfaa8430, 0xc000392960, 0x61b4340, 0xc0009f5470, 0x5cd5e40, 0xc000ae3f50, 0xc0002e0400, 0xfb82290)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/internal/go_templates/texttemplate/hugo_template.go:79 +0x1eb fp=0xc0008bf860 sp=0xc0008bf780 pc=0x4a8a40b
github.com/gohugoio/hugo/tpl/tplimpl.(*templateExec).Execute(0xc000ae3050, 0x61c8600, 0xc000392960, 0x61b4340, 0xc0009f5470, 0x5cd5e40, 0xc000ae3f50, 0x0, 0x0)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/tpl/tplimpl/template.go:227 +0x168 fp=0xc0008bf930 sp=0xc0008bf860 pc=0x4d806a8
github.com/gohugoio/hugo/hugolib.(*Site).renderForTemplate(0xc0000d78c0, 0x5ce06a5, 0x4, 0xc000805e68, 0x3, 0x5cd5e40, 0xc000ae3f50, 0x61b4340, 0xc0009f5470, 0x61c8600, ...)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/hugolib/site.go:1759 +0x96 fp=0xc0008bf9a8 sp=0xc0008bf930 pc=0x4dd9356
github.com/gohugoio/hugo/hugolib.(*Site).renderAndWritePage(0xc0000d78c0, 0xc0002165b0, 0xc000b88068, 0x8, 0xc000adcf60, 0xe, 0xc000ae3f50, 0x61c8600, 0xc000392960, 0x0, ...)
        /private/tmp/hugo-20201113-76036-iuot5h/hugo-0.78.2/src/github.com/gohugoio/hugo/hugolib/site.go:1692 +0x2e6 fp=0xc0008bfea0 sp=0xc0008bf9a8 pc=0x4dd8c06
github.com/gohugoio/hugo/hugolib.pageRenderer(0xc000803fa0, 0xc0000d78c0, 0xc0008bb0e0, 0xc00009f560, 0xc00061aca0)
```

**Comment by jmooring** [comment]
@moorereason Was this resolved with #8149?

**Comment by bep** [comment]
I'm pretty sure this is fixed.
